internal/bg/availability_queue.go

```diff
@@ -0,0 +1,64 @@
+package bg
+
+import (
+	"context"
+	"time"
+
+	"github.com/RHEnVision/provisioning-backend/internal/bus"
+	"github.com/RHEnVision/provisioning-backend/internal/ctxval"
+)
+
+// buffered channel for incoming requests
+var asRequestChan = make(chan *bus.GenericMessage, availabilityStatusBatchSize)
+
+// EnqueueAvailabilityStatusRequest prepares a status request check to be sent in the next
+// batch to the platform bus. Messages can be delayed up to several seconds until sent.
+// The function can block if the enqueueing channel is full.
+func EnqueueAvailabilityStatusRequest(msg *bus.GenericMessage) {
+	asRequestChan <- msg
+}
+
+func send(ctx context.Context, messages ...*bus.GenericMessage) {
+	err := bus.Send(ctx, messages...)
+	if err != nil {
+		ctxval.Logger(ctx).Warn().Err(err).Msg("Unable to send availability check messages")
+	}
+}
+
+// main sending loop: takes messages enqueued via EnqueueAvailabilityStatusRequest and sends them to the bus
+func sendAvailabilityRequestMessages(ctx context.Context, batchSize int, tickDuration time.Duration) {
+	logger := ctxval.Logger(ctx)
+	ticker := time.NewTicker(tickDuration)
+	messageBuffer := make([]*bus.GenericMessage, 0, batchSize)
+
+	for {
+		select {
+		case msg := <-asRequestChan:
+			messageBuffer = append(messageBuffer, msg)
+			length := len(messageBuffer)
+
+			if length >= batchSize {
+				logger.Trace().Int("messages", length).Msgf("Sent %d availability request messages (full buffer)", length)
+				send(ctx, messageBuffer...)
+				messageBuffer = make([]*bus.GenericMessage, 0, batchSize)
+			}
+		case <-ticker.C:
+			length := len(messageBuffer)
+
+			if length > 0 {
+				logger.Trace().Int("messages", length).Msgf("Sent %d availability request messages (tick)", length)
+				send(ctx, messageBuffer...)
+				messageBuffer = make([]*bus.GenericMessage, 0, batchSize)
+			}
+		case <-ctx.Done():
+			ticker.Stop()
+			length := len(messageBuffer)
+
+			if length > 0 {
+				logger.Trace().Int("messages", length).Msgf("Sent %d availability request messages (cancel)", length)
+				send(ctx, messageBuffer...)
+				messageBuffer = make([]*bus.GenericMessage, 0, batchSize)
```

Comment:
Ha these three reallocations need to be reslicing instead, this creates unnecessary GC pressure. Will fix this in the next rebase.

				messageBuffer = messageBuffer[:0]